trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/library/PthreadLibraryModel.java

```diff
@@ -0,0 +1,500 @@
+
+		return handleLockCall(main, loc, name, lock, callFactory);
+	}
+
+	private ExpressionResult handleLockCall(final IDispatcher main, final ILocation loc, final String name,
+			final IASTInitializerClause lock, final ILockCallFactory callFactory) {
+		final ExpressionResultBuilder erb = new ExpressionResultBuilder();
+
+		final ExpressionResult arg = mExprResultTransformer.transformDispatchDecaySwitchRexBoolToInt(main, loc, lock);
+		final Expression index = arg.getLrValue().getValue();
+		erb.addAllExceptLrValue(arg);
+
+		// auxvar for procedure's return value
+		final ICType cType = new CPrimitive(CPrimitives.INT);
+		final AuxVarInfo auxvarinfo = mAuxVarInfoBuilder.constructAuxVarInfo(loc, cType, SFO.AUXVAR.RETURNED);
+		erb.addAuxVarWithDeclaration(auxvarinfo);
+
+		erb.addStatement(callFactory.apply(loc, index, auxvarinfo.getLhs()));
+		erb.setLrValue(new RValue(auxvarinfo.getExp(), new CPrimitive(CPrimitives.INT)));
+		return erb.build();
+	}
+
+	private interface ILockCallFactory {
+		Statement apply(ILocation loc, Expression index, VariableLHS lhs);
+	}
+
+	private Result handlePthread_mutex_init(final IDispatcher main, final IASTFunctionCallExpression node,
+			final ILocation loc, final String name) {
+		mMemoryHandler.requireMemoryModelFeature(MemoryModelDeclarations.ULTIMATE_PTHREADS_MUTEX);
+
+		final IASTInitializerClause[] arguments = node.getArguments();
+		mHelper.checkArguments(loc, 2, name, arguments);
+
+		final ExpressionResult arg1 =
+				mExprResultTransformer.transformDispatchDecaySwitchRexBoolToInt(main, loc, arguments[0]);
+		final ExpressionResult arg2 =
+				mExprResultTransformer.transformDispatchDecaySwitchRexBoolToInt(main, loc, arguments[1]);
+		final boolean isNullPointerLiteral = mMemoryHandler.isNullPointerLiteral(arg2.getLrValue().getValue());
+		if (!isNullPointerLiteral) {
+			final String msg = "The second argument of the pthread_mutex_init is not a null pointer. This means that "
+					+ "non-default attributes are used. We support only the default attributes.";
+			throw new UnsupportedSyntaxException(loc, msg);
+		}
+
+		final CPrimitive returnType = new CPrimitive(CPrimitives.INT);
+		// we assume that function is always successful and returns 0
+		final BigInteger value = BigInteger.ZERO;
+		final Expression index = arg1.getLrValue().getValue();
+		final AssignmentStatement unlockMutex = mMemoryHandler.constructMutexArrayAssignment(loc, index, false);
+		final ExpressionResultBuilder erb = new ExpressionResultBuilder();
+		erb.addAllExceptLrValue(arg1);
+		erb.addStatement(unlockMutex);
+		erb.setLrValue(new RValue(mTypeSizes.constructLiteralForIntegerType(loc, returnType, value),
+				new CPrimitive(CPrimitives.INT)));
+		return erb.build();
+	}
+
+	private TypesResult handlePthread_t(final IASTNamedTypeSpecifier node, final ILocation loc) {
+		final var cType = mTypeHandler.getThreadIdType();
+		return new TypesResult(mTypeHandler.cType2AstType(loc, cType), node.isConst(), false, cType);
```

Comment:
Currently, `TypeModel` is quite similar to `FunctionModel`: It contains a name and a function to produce a `Result` (should be always a `TypesResult`). But it looks like it could be simplified. All those function seem to produce some `CType` `cType` and return `new TypesResult(mTypeHandler.cType2AstType(loc, cType), node.isConst(), false, cType)`. If we don't need this flexibility to use a function, we could also directly use a `CType` in `TypeModel` instead.

Reply:
So, I just tried replacing the functional interface with a simple `ICType` (see 89341c3). This simplifies the code, but we cannot define types with a custom `ASTType`, but I am not sure if we need that either. This removed also the bug above with `BoogieType.TYPE_REAL`.